### PR TITLE
CB-12021 : --fetch should use dependency version saved in package.json before default cordova version when adding/restoring

### PR DESCRIPTION
**Pull request by audreyso**
### What does this PR do?
 --fetch should use dependency version saved in package.json before default cordova version when adding/restoring

### Checklist
- [X] [Reported an issue](http://cordova.apache.org/contribute/issues.html) in the JIRA database
- [X] Commit message follows the format: "CB-3232: (android) Fix bug with resolving file paths", where CB-xxxx is the JIRA ID & "android" is the platform affected.
- [X] Added automated test coverage as appropriate for this change.

### CODE REVIEW

File: cordova-lib/src/plugman/fetch.js [hunk 1]
```diff
@@ -69,7 +71,7 @@ function fetchPlugin(plugin_src, plugins_dir, options) {
                 options.subdir = result[2];
             //if --fetch was used, throw error for subdirectories
 
-            if (options.subdir && options.subdir !== '.') {
```
Comment (stevengill) [inline cordova-lib/src/plugman/fetch.js]: I think this may have been taken out by mistake

File: cordova-lib/src/cordova/platform.js [hunk 2]
```diff
@@ -253,6 +275,7 @@ function addHelper(cmd, hooksRunner, projectRoot, targets, opts) {
             if(fs.existsSync(pkgJsonPath)) {
                 delete require.cache[require.resolve(pkgJsonPath)]; 
                 pkgJson = require(pkgJsonPath);
+
```
Comment (stevengill) [inline cordova-lib/src/cordova/platform.js]: nit: unnecessary line

File: cordova-lib/src/cordova/platform.js [hunk 3]
```diff
@@ -120,13 +131,22 @@ function addHelper(cmd, hooksRunner, projectRoot, targets, opts) {
...
+                        } else {
+                            return getPlatformDetailsFromDir(maybeDir, platform);
+                        }
                     }
                 }
                 return downloadPlatform(projectRoot, platform, spec, opts);
             }).then(function(platDetails) {
                 if(fs.existsSync(path.join(projectRoot, 'package.json'))) {
                     delete require.cache[require.resolve(path.join(projectRoot, 'package.json'))];
+                    var pkgJson;
```
Comment (stevengill) [inline cordova-lib/src/cordova/platform.js]: i don't think 148-149 are used

File: cordova-lib/spec-cordova/platform.spec.js [hunk 4]
```diff
@@ -353,4 +353,5 @@ describe('plugin add and rm end-to-end --fetch', function () {
         })
         .fin(done);
     }, 60000);
-});
\ No newline at end of file
+});
+
```
Comment (stevengill) [inline cordova-lib/spec-cordova/platform.spec.js]: nit: delete extra new line

File: cordova-lib/spec-cordova/pkgJson-restore.spec.js [hunk 5]
```diff
@@ -742,7 +742,7 @@ describe('update config.xml to use the variable found in pkg.json', function ()
         process.chdir(path.join(__dirname, '..'));  // Needed to rm the dir on Windows.
         shell.rm('-rf', tmpDir);
     });
-
+
```
Comment (stevengill) [inline cordova-lib/spec-cordova/pkgJson-restore.spec.js]: nit: unnecessary tab

File: cordova-lib/package.json [hunk 6]
```diff
@@ -55,7 +55,7 @@
     "test": "npm run jshint && npm run jasmine",
     "test-ios": "npm run test && npm run jasmine-ios",
     "ci": "npm run jshint && npm run cover && codecov",
-    "jshint": "jshint src spec-cordova spec-plugman",
+    "jshint": "./node_modules/.bin/jshint src spec-cordova spec-plugman",
```
Comment (stevengill) [inline cordova-lib/package.json]: not sure if full path to jshint is needed anymore. That error you were getting earlier was solved by deleting browser fixture's tests

File: cordova-lib/src/cordova/plugin.js [hunk 7]
```diff
@@ -388,26 +384,54 @@ module.exports = function plugin(command, targets, opts) {
...
+
+    // If parsedSpec.version satisfies pkgJson version, no writing to pkg.json. Only write when
+    // it does not satisfy.
+    if(parsedSpec.version) {
+        if(pkgJson && pkgJson.dependencies && pkgJson.dependencies[parsedSpec.package]) {
+            var noSymbolVersion;
+            if (parsedSpec.version.charAt(0) === '^' || parsedSpec.version.charAt(0) === '~') {
+                noSymbolVersion = parsedSpec.version.slice(1);
+            }
+            if (!semver.satisfies(noSymbolVersion, pkgJson.dependencies[parsedSpec.package])) {
```
Comment (audreyso) [inline cordova-lib/src/cordova/plugin.js]: How does this look?

   ```// If parsedSpec.version satisfies pkgJson version, no writing to pkg.json. Only write when
    // it does not satisfy.
    if(parsedSpec.version) {
        if(pkgJson && pkgJson.dependencies && pkgJson.dependencies[parsedSpec.package]) {
            var noSymbolVersion = parsedSpec.version;
            if (parsedSpec.version.charAt(0) === '^' || parsedSpec.version.charAt(0) === '~') {
                noSymbolVersion = parsedSpec.version.slice(1);
            }
            if (pkgJson.dependencies[parsedSpec.package] === parsedSpec.version) {
                pkgJson.dependencies[parsedSpec.package] = parsedSpec.version;
                if (fetchOptions.save === true) {
                    fs.writeFileSync(pkgJsonPath, JSON.stringify(pkgJson, null, 4), 'utf8');
                }
            }
        }
    }

    if (cordova_util.isUrl(parsedSpec.version) || cordova_util.isDirectory(parsedSpec.version) || pluginSpec.parse(parsedSpec.version).scope) {
        return Q(parsedSpec.version);
    }```

File: cordova-lib/src/cordova/plugin.js [hunk 8]
```diff
@@ -414,13 +414,15 @@ function determinePluginTarget(projectRoot, cfg, target, fetchOptions) {
...
-            var noSymbolVersion;
+            var noSymbolVersion = parsedSpec.version;
             if (parsedSpec.version.charAt(0) === '^' || parsedSpec.version.charAt(0) === '~') {
                 noSymbolVersion = parsedSpec.version.slice(1);
             }
-            if (!semver.satisfies(noSymbolVersion, pkgJson.dependencies[parsedSpec.package])) {
+            if (pkgJson.dependencies[parsedSpec.package] === parsedSpec.version) {
                 pkgJson.dependencies[parsedSpec.package] = parsedSpec.version;
-                fs.writeFileSync(pkgJsonPath, JSON.stringify(pkgJson, null, 4), 'utf8');
+                if (fetchOptions.save === true) {
```
Comment (stevengill) [inline cordova-lib/src/cordova/plugin.js]: this is good!

File: cordova-lib/src/cordova/plugin.js [hunk 9]
```diff
@@ -414,13 +414,15 @@ function determinePluginTarget(projectRoot, cfg, target, fetchOptions) {
     // it does not satisfy.
     if(parsedSpec.version) {
         if(pkgJson && pkgJson.dependencies && pkgJson.dependencies[parsedSpec.package]) {
-            var noSymbolVersion;
+            var noSymbolVersion = parsedSpec.version;
             if (parsedSpec.version.charAt(0) === '^' || parsedSpec.version.charAt(0) === '~') {
                 noSymbolVersion = parsedSpec.version.slice(1);
             }
-            if (!semver.satisfies(noSymbolVersion, pkgJson.dependencies[parsedSpec.package])) {
+            if (pkgJson.dependencies[parsedSpec.package] === parsedSpec.version) {
```
Comment (stevengill) [inline cordova-lib/src/cordova/plugin.js]: I think we still want to do the semver check too. First start with outer if that checks if `parsedSpec.version` is a directory or url,  if it is, do this check  to see if `pkgJson.dependencies[parsedSpec.package] !== parsedSpec.version`. If that is true, then you want to update package.json (assuming fetchOpts.save === true)

On that outer if that checks if it is a dir or url, add a else if where you check for semver.satisfies

File: cordova-lib/src/cordova/plugin.js [hunk 10]
```diff
@@ -412,15 +412,26 @@ function determinePluginTarget(projectRoot, cfg, target, fetchOptions) {
...
     if(parsedSpec.version) {
         if(pkgJson && pkgJson.dependencies && pkgJson.dependencies[parsedSpec.package]) {
-            var noSymbolVersion;
+            var noSymbolVersion = parsedSpec.version;
             if (parsedSpec.version.charAt(0) === '^' || parsedSpec.version.charAt(0) === '~') {
                 noSymbolVersion = parsedSpec.version.slice(1);
             }
-            if (!semver.satisfies(noSymbolVersion, pkgJson.dependencies[parsedSpec.package])) {
+
+            if(cordova_util.isUrl(parsedSpec.version) || cordova_util.isDirectory(parsedSpec.version)) {
```
Comment (audreyso) [inline cordova-lib/src/cordova/plugin.js]: How does this look (lines 423- 437)? @stevengill

Comment (stevengill) [inline cordova-lib/src/cordova/plugin.js]: LGTM!

**Comment by stevengill** [comment]
I think this is ready to merge in

**Pull request closed without merging**